File: 0109-convert-sorted-list-to-binary-search-tree/0109-convert-sorted-list-to-binary-search-tree.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
# 自定义迭代器类
class ArrayIterator:
    def __init__(self, nums):
        self.nums = nums
        self.index = 0
    
    def next(self):
        if self.index < len(self.nums):
            value = self.nums[self.index]
            self.index += 1
            return value
        return None  # 如果没有更多元素，返回 None
    
    def hasNext(self):
        return self.index < len(self.nums)
class Solution:
    def sortedListToBST(self, head: Optional[ListNode]) -> Optional[TreeNode]:
        # Step 1: 将链表转换为有序数组
        nums = []
        while head:
            nums.append(head.val)
            head = head.next
        
        # Step 2: 创建自定义迭代器
        nums_iter = ArrayIterator(nums)
        
        # Step 3: 使用自定义迭代器递归构建平衡二叉搜索树
        def buildTree(nums_iter, start, end):
            if start > end:
                return None
            
            mid = (start + end) // 2
            node_val = nums_iter.nums[mid]  # 获取中间值
            node = TreeNode(node_val)
            
            node.left = buildTree(nums_iter, start, mid - 1)  # 递归构建左子树
            node.right = buildTree(nums_iter, mid + 1, end)   # 递归构建右子树
            return node
        
        # Step 4: 返回构建好的树的根节点
        return buildTree(nums_iter, 0, len(nums) - 1)

# 作者：LisaHusband
# 链接：https://leetcode.cn/problems/convert-sorted-list-to-binary-search-tree/solutions/
# 来源：力扣（LeetCode）
# 著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。

File: 0109-convert-sorted-list-to-binary-search-tree/0109-convert-sorted-list-to-binary-search-tree.go
package main

import "fmt"


// 自定义迭代器类
type ArrayIterator struct {
    nums  []int
    index int
}

func (it *ArrayIterator) next() int {
    if it.index < len(it.nums) {
        value := it.nums[it.index]
        it.index++
        return value
    }
    return -1 // 如果没有更多元素，返回 -1
}

func (it *ArrayIterator) hasNext() bool {
    return it.index < len(it.nums)
}

// Solution: 将有序链表转换为平衡二叉搜索树
func sortedListToBST(head *ListNode) *TreeNode {
    // Step 1: 将链表转换为有序数组
    nums := []int{}
    for head != nil {
        nums = append(nums, head.Val)
        head = head.Next
    }

    // Step 2: 创建自定义迭代器
    numsIter := ArrayIterator{nums: nums}

    // Step 3: 使用自定义迭代器递归构建平衡二叉搜索树
    return buildTree(&numsIter, 0, len(nums)-1)
}

func buildTree(numsIter *ArrayIterator, start, end int) *TreeNode {
    if start > end {
        return nil
    }

    mid := (start + end) / 2
    nodeVal := numsIter.nums[mid]
    node := &TreeNode{Val: nodeVal}

    node.Left = buildTree(numsIter, start, mid-1)  // 递归构建左子树
    node.Right = buildTree(numsIter, mid+1, end)   // 递归构建右子树

    return node
}

// Helper function to print the tree (inorder traversal)
func inorderTraversal(root *TreeNode) {
    if root != nil {
        inorderTraversal(root.Left)
        fmt.Printf("%d ", root.Val)
        inorderTraversal(root.Right)
    }
}


